File: src/main/java/macro303/keschet/board/Board.java
package macro303.keschet.board;

import macro303.keschet.Colour;
import macro303.keschet.Coordinates;
import macro303.keschet.Util;
import macro303.keschet.pieces.Piece;
import macro303.keschet.players.Player;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Created by Macro303 on 2018-02-08.
 */
public class Board {
	private static final Logger LOGGER = LogManager.getLogger(Board.class);
	@NotNull
	private final Square[][] board = new Square[Util.SIZE][Util.SIZE];

	public Board() {
		for (int row = 0; row < Util.SIZE; row++)
			for (int col = 0; col < Util.SIZE; col++)
				board[col][row] = new Square(new Coordinates(row, col));
	}

	@Nullable
	public Square getSquare(@NotNull Coordinates location) {
		if (location.getRow() < 0 || location.getRow() >= Util.SIZE || location.getCol() < 0 || location.getCol() >= Util.SIZE)
			return null;
		return board[location.getCol()][location.getRow()];
	}

	public int countPieces(@NotNull Player player) {
		int counter = 0;
		for (int row = 0; row < Util.SIZE; row++) {
			for (int col = 0; col < Util.SIZE; col++) {
				Square location = getSquare(new Coordinates(row, col));
				assert location != null;
				if (location.getPiece() != null && location.getPiece().getTeamColour() == player.getTeamColour())
					counter++;
			}
		}
		return counter;
	}

	@Nullable
	public Square findPiece(@NotNull Class clazz, @NotNull Colour teamColour) {
		Square temp = null;
		for (int row = 0; row < Util.SIZE; row++) {
			for (int col = 0; col < Util.SIZE; col++) {
				Square location = getSquare(new Coordinates(row, col));
				assert location != null;
				if (location.getPiece() != null && location.getPiece().getClass().equals(clazz) && location.getPiece().getTeamColour() == teamColour)
					temp = location;
			}
		}
		return temp;
	}
}

File: Base/src/main/kotlin/macro303/base/Board.kt
package macro303.base

import macro303.base.pieces.Emperor
import macro303.base.pieces.Piece
import org.apache.logging.log4j.LogManager
import java.util.*

/**
 * Created by Macro303 on 2017-10-20.
 */
class Board : IBoard {
	private var board = Array(SIZE, { Array(SIZE, { Square() }) })

	fun draw() {
		(0..SIZE).forEach { column ->
			(0..SIZE).forEach {
				when {
					column == 0 -> print("${if (it == 0) "" else " "}$it${if (it == SIZE) "" else " "}")
					it == 0 -> print("$column${if (column == SIZE) "" else " "}")
					else -> Console.showSquare(getSquare(Pair(it, column))!!.piece)
				}
			}
			println()
		}
	}

	override fun getSquare(square: Pair<Int, Int>): Square? {
		return if (square.first - 1 < 0 || square.first > SIZE || square.second - 1 < 0 || square.second > SIZE) null else board[square.second - 1][square.first - 1]
	}

	override fun countPieces(team: Team): Int {
		var counter = 0
		board.forEach { it -> counter += it.count { it.piece != null && it.piece?.teamColour == team.colour && it.piece !is Emperor } }
		LOGGER.trace("int countPieces(Player) = $counter")
		return counter
	}

	override fun pieceStillOnBoard(clazz: Class<*>, teamColour: Colour): Boolean {
		val stillAvailable =
			board.any { it.any { it.piece != null && it.piece?.teamColour == teamColour && it.piece?.javaClass == clazz } }
		LOGGER.trace("boolean pieceStillOnBoard(Class, Colour) = $stillAvailable")
		return stillAvailable
	}

	override fun getAllSurroundingPieces(squareLocation: Pair<Int, Int>): ArrayList<Piece> {
		val pieces = ArrayList<Piece>()
		var square = getSquare(square = Pair(squareLocation.first, squareLocation.second + 1))
		if (square != null && square.piece != null)
			pieces.add(square.piece!!)
		square = getSquare(square = Pair(squareLocation.first + 1, squareLocation.second + 1))
		if (square != null && square.piece != null)
			pieces.add(square.piece!!)
		square = getSquare(square = Pair(squareLocation.first + 1, squareLocation.second))
		if (square != null && square.piece != null)
			pieces.add(square.piece!!)
		square = getSquare(square = Pair(squareLocation.first + 1, squareLocation.second - 1))
		if (square != null && square.piece != null)
			pieces.add(square.piece!!)
		square = getSquare(square = Pair(squareLocation.first, squareLocation.second - 1))
		if (square != null && square.piece != null)
			pieces.add(square.piece!!)
		square = getSquare(square = Pair(squareLocation.first - 1, squareLocation.second - 1))
		if (square != null && square.piece != null)
			pieces.add(square.piece!!)
		square = getSquare(square = Pair(squareLocation.first - 1, squareLocation.second))
		if (square != null && square.piece != null)
			pieces.add(square.piece!!)
		square = getSquare(square = Pair(squareLocation.first - 1, squareLocation.second + 1))
		if (square != null && square.piece != null)
			pieces.add(square.piece!!)
		LOGGER.trace("ArrayList<Piece> getAllSurroundingPieces(Pair<Int, Int>) = $pieces")
		return pieces
	}

	companion object {
		private val LOGGER = LogManager.getLogger(Board::class.java)
		private const val SIZE = 10
	}
}